nomenklatura/index/duckdb_index.py

```diff
@@ -0,0 +1,162 @@
+from duckdb import DuckDBPyRelation
+from followthemoney.types import registry
+from pathlib import Path
+from shutil import rmtree
+from typing import Iterable, Tuple
+import csv
+import duckdb
+import logging
+
+from nomenklatura.dataset import DS
+from nomenklatura.entity import CE
+from nomenklatura.index.common import BaseIndex
+from nomenklatura.index.tokenizer import NAME_PART_FIELD, WORD_FIELD, Tokenizer
+from nomenklatura.resolver import Pair, Identifier
+from nomenklatura.store import View
+
+log = logging.getLogger(__name__)
+
+BATCH_SIZE = 1000
+
+
+class DuckDBIndex(BaseIndex[DS, CE]):
+    """
+    An in-memory search index to match entities against a given dataset.
+
+    For each field in the dataset, the index stores the IDs which contains each
+    token, along with the absolute frequency of each token in the document.
+    """
+
+    name = "duckdb"
+
+    BOOSTS = {
+        NAME_PART_FIELD: 2.0,
+        WORD_FIELD: 0.5,
+        registry.name.name: 10.0,
+        # registry.country.name: 1.5,
+        # registry.date.name: 1.5,
+        # registry.language: 0.7,
+        # registry.iban.name: 3.0,
+        registry.phone.name: 3.0,
+        registry.email.name: 3.0,
+        # registry.entity: 0.0,
+        # registry.topic: 2.1,
+        registry.address.name: 2.5,
+        registry.identifier.name: 3.0,
+    }
+
+    __slots__ = "view", "fields", "tokenizer", "entities"
+
+    def __init__(self, view: View[DS, CE], data_dir: Path):
+        self.view = view
+        self.tokenizer = Tokenizer[DS, CE]()
+        self.data_dir = data_dir
+        if self.data_dir.exists():
+            rmtree(self.data_dir.as_posix())
+        self.data_dir.mkdir(parents=True)
+        self.con = duckdb.connect((self.data_dir / "duckdb_index.db").as_posix())
+
+        # https://duckdb.org/docs/guides/performance/environment
+        # > For ideal performance, aggregation-heavy workloads require approx.
+        # > 5 GB memory per thread and join-heavy workloads require approximately
+        # > 10 GB memory per thread.
+        # > Aim for 5-10 GB memory per thread.
+        self.con.execute("SET memory_limit = '2GB';")
+        self.con.execute("SET max_memory = '2GB';")
+        # > If you have a limited amount of memory, try to limit the number of threads
+        self.con.execute("SET threads = 1;")
```

Comment:
this can probably be improved on, but this is working with `default` with a 10GB container memory limit.

Configuring much less memory here resulted in duckdb out of memory exception. Configuring more resulted in the process getting OOMKilled with a 10GB limit. Changing pairs to a generator probably gave more headroom  but each iteration takes 7 minutes to test.